homeassistant/components/yolink/switch.py

```diff
@@ -108,18 +155,32 @@ def __init__(
             f"{coordinator.device.device_name} ({self.entity_description.name})"
         )
 
+    def _get_state(
+        self, state_value: str | list[str] | None, plug_index: int | None
+    ) -> bool | None:
+        """Parse state value."""
+        if isinstance(state_value, list) and plug_index is not None:
+            return state_value[plug_index] == "open"
+        return state_value == "open" if state_value is not None else None
+
     @callback
-    def update_entity_state(self, state: dict[str, Any]) -> None:
+    def update_entity_state(self, state: dict[str, str | list[str]]) -> None:
         """Update HA Entity State."""
-        self._attr_is_on = self.entity_description.value(
-            state.get(self.entity_description.state_key)
+        self._attr_is_on = self._get_state(
+            state.get("state"), self.entity_description.plug_index
         )
         self.async_write_ha_state()
 
     async def call_state_change(self, state: str) -> None:
         """Call setState api to change switch state."""
-        await self.call_device_api("setState", {"state": state})
-        self._attr_is_on = self.entity_description.value(state)
+        await self.call_device(
+            OutletReqeustBuilder()
+            .method("setState")
```

Comment:
I'd recommend defining each method instead so that the library user can easily know the available api and have a typed interface for each method.

Reply:
like this ?
```
class OutletReqeustBuilder:
    """Outlet request builder"""

    @classmethod
    def set_state_reqeust(cls, state: str, plug_indx: int | None) -> ClientReqeust:
        """Set device state."""
        params: dict[str, str | int] = {"state": state}
        if plug_indx is not None:
            params["chs"] = 1 << plug_indx
        return ClientReqeust("setState", params)
```

```
async def call_state_change(self, state: str) -> None:
        """Call setState api to change switch state."""
        await self.call_device(
            OutletReqeustBuilder.set_state_reqeust(state, self.entity_description.plug_index)
        )
        self._attr_is_on = self._get_state(state, self.entity_description.plug_index)
        self.async_write_ha_state()

```

Reply:
Yes, that's better. Note the typo on request.